src/sagemaker/local/image.py

```diff
@@ -0,0 +1,508 @@
+        for host in self.hosts:
+            _create_config_file_directories(self.container_root, host)
+            self.write_config_files(host, hyperparameters, input_data_config)
+            shutil.copytree(data_dir, os.path.join(self.container_root, host, 'input', 'data'))
+
+        compose_data = self._generate_compose_file('train', additional_volumes=volumes)
+        compose_command = self._compose()
+        _execute_and_stream_output(compose_command)
+
+        s3_model_artifacts = self.retrieve_model_artifacts(compose_data)
+
+        # free up the training data directory as it may contain
+        # lots of data downloaded from S3. This doesn't delete any local
+        # data that was just mounted to the container.
+        shutil.rmtree(data_dir)
+        # Also free the container config files.
+        for host in self.hosts:
+            shutil.rmtree(os.path.join(self.container_root, host))
+
+        self._cleanup()
+        return s3_model_artifacts
+
+    def serve(self, primary_container):
+        """Host a local endpoint using docker-compose.
+        Args:
+            primary_container (dict): dictionary containing the container runtime settings
+                for serving. Expected keys:
+                - 'ModelDataUrl' pointing to a local file
+                - 'Environment' a dictionary of environment variables to be passed to the hosting container.
+
+        """
+        logger.info("serving")
+
+        self.container_root = self._create_tmp_folder()
+        logger.info('creating hosting dir in {}'.format(self.container_root))
+
+        model_dir = primary_container['ModelDataUrl']
+        if not model_dir.lower().startswith("s3://"):
+            for h in self.hosts:
+                host_dir = os.path.join(self.container_root, h)
+                os.makedirs(host_dir)
+                shutil.copytree(model_dir, os.path.join(self.container_root, h, 'model'))
+
+        env_vars = ['{}={}'.format(k, v) for k, v in primary_container['Environment'].items()]
+
+        self._generate_compose_file('serve', additional_env_vars=env_vars)
+        compose_command = self._compose()
+        self.container = _HostingContainer(compose_command)
+        self.container.up()
+
+    def stop_serving(self):
+        """Stop the serving container.
+
+        The serving container runs in async mode to allow the SDK to do other tasks.
+        """
+        if self.container:
+            self.container.down()
+            self._cleanup()
+        # for serving we can delete everything in the container root.
+        shutil.rmtree(self.container_root)
+
+    def retrieve_model_artifacts(self, compose_data):
+        """Get the model artifacts from all the container nodes.
+
+        Used after training completes to gather the data from all the individual containers. As the
+        official SageMaker Training Service, it will override duplicate files if multiple containers have
+        the same file names.
```

Comment:
I believe that we should save the model in the S3 bucket in the same way that the model is saved in SageMaker Training Service. My reasoning is that other services in the future may rely with the model in S3.
We can provide methods to allow the customer to easily download the models but that is a nice to have.